Hep_methylation_TF
===
Last update: 20212-03-08

Author: Takahiro Suzuki (takahiro.suzuki.aa@riken.jp)
<br><br>

## **Description**
## src/shell
---
### 1. beam2peaks_ATAC.sh

shell script for ChIPmentation peakcalling. 

First, from a bam file, remove the mitochondrial genome, duplicated, and "blacklist" reads.
Then, perform peakcalling with MACS2.

Usage:
```bash
 bash ./src/shell/bam2peaks_ATAC.sh -f [bam file] -r s -o [prefix] -p 1e-5
 ```
processed bam files and a directory of MACS2 result are generated.


### 2. bam2peaks_ChIP.sh
shell script for ChIPmentation peakcalling. 

First, from a bam file, remove the mitochondrial genome, duplicated, and "blacklist" reads.
Then, perform peakcalling with MACS2.

Usage:
```bash
 bash ./src/shell/bam2peaks_ChIP.sh -f [bam file] -r p -o [prefix] -p 1e-10
 ```
processed bam files and a directory of MACS2 result are generated.

<br><br>
For the preocessed bam file, bigwig files can be generated by following script.
```bash
bam2wig.py -t 1000000000 -i [bam file] -o [prefix] -u -s ./data/hg19.chrom.sizes
```
Dependency:

bam2wig.py (RseQC: http://rseqc.sourceforge.net/)

wigToBigWig (https://www.encodeproject.org/software/wigtobigwig/)



<br><br>

## src/R
---
### 1. Methyl850_HEP.r
- Preprocessing

   Normalization and M-value computation

- correlation matrix (Fg. 1f)
- DMR b2b plot (Fig. 1g)
- Functional analysis of DMRs (Supplementary Figure 3)
- Heatmap for methylation regulating TF (Fig. 2b and 2d)
- exprssion line plot (Fig. 2c and 2e; Supplementary Figure 4)

### 2. CAGE_HEP.r
- Expression Table for For  "Heatmap for methylation regulating TF (Fig. 2b, d)" in Methyl850_HEP.r
- Marker gene exprssion (Supplementary Figure 1c and 2c)

   Plots showing GOI expression pattern of GOIs


### 4. Methyl850_DE.r
- preprocessing 

   Normalization and M-value computation
   
- DMR b2b plot (Fig. 4b)
- enrichment plot (Supplementary Figure 6b and 6c)
- overlap of demethylated probes(upset plot)(Fig. 4c)
- methylation heatmap for uninheited demet probes (Supplementary Figure 6d)
- enrichment plot for uninherited demethylated regions (Fig. 4d)
- Enrichment heatmap(Fig. 4e and 5b)
- expression analysis for demet probes (Fig. 5a)
- Functional Annotation of DMRs (Supplementary Figure 7a)

### 5. CAGE_DE.r (Supplementary Figure 6a)

### 6. multi-omics.r(Fig. 6a)

### 7. ATAC_DE.r (Supplementary Figure 7c)

### 8. OE_DE_HEP_intersect.r (Supplementary Figure 6e)

<br><br>

## data/
---
- CAGE_HEP/gene_expMatrix.RData

R data containg a hepatocyte time-course expression matrix

- Methyl850_HEP/
    - motif analysis/
    - Hep_00_07
    - Hep_07_14
    - Hep_14_21
    - Hep_21_28

These directories contain results of motif overrepresentation analyses for demethylated regions (Demethy_mot_analysis_result.txt) and for methylated regions(Methyl_mot_analysis result.txt).

- Demethyl_motif_hatmap.RData and Methyl_motif_heatmap.RData

R data contating all objects in the Motf overrepresenation analysis

- Methyl850_DE/
    - bed_d1/*:   bed files of demethylated CpGs
    - bed_d500/*:   bed files of +- 250 bp region from demethylated CpGs
    - exp_table/*:   transcripts expression tables of demethylated regions (+- 250 bp region from demethylated CpGs) 
    
- genecode.v19.gene_only.gff3

The gencode human v19 data containing genes only

- human_permissive_enhancers_phase1_and_2.bed

The bed file of FANTOM5 permissive enhancers 

- hg19_blacklist.bed
The ENCODE blacklist

- hg19.chrom.sizes
Chromosome size for bigwig

- MethylationEPICv-1-0.B4.bed
Methylation array annotation file